druntime/src/rt/sections_win64.d

```diff
@@ -104,19 +133,80 @@ void initSections() nothrow @nogc
                 ranges[r++] = (cast(void**)addr)[0..1];
             prev = addr;
         }
-        _sections._gcRanges = ranges[0..r];
+        sectionGroup._gcRanges = ranges[0..r];
+        sectionGroup._tpSection = findImageSection(handle, ".tp");
     }
+    _sections.insertBack(sectionGroup);
 }
 
 /***
  * Gets called on program shutdown just after GC is terminated.
  */
 void finiSections() nothrow @nogc
 {
-    .free(cast(void*)_sections.modules.ptr);
-    .free(_sections._gcRanges.ptr);
+    foreach_reverse (ref sec; _sections)
+        finiSections(sec);
+    _sections.reset();
+}
+
+void finiSections(SectionGroup* sec) nothrow @nogc
+{
+    .free(cast(void*)sec.modules.ptr);
+    .free(sec._gcRanges.ptr);
+    .free(sec);
 }
 
+private void scanTLSPrecise(const(uint)* tp_beg, const(uint)* tp_end, void* base,
+                            scope void delegate(void* pbeg, void* pend) nothrow dg) nothrow
+{
+    for (auto p = tp_beg; p < tp_end; )
+    {
+        uint beg = *p++;
+        uint end = beg + cast(uint)((void*).sizeof);
+        while (p < tp_end && *p == end)
+        {
+            end += (void*).sizeof;
+            p++;
+        }
+        dg(base + beg, base + end);
+    }
+}
+
+version (Shared)
+{
+    void** initTLSRanges() nothrow @nogc
+    {
+        return getTEB();
+    }
+    void finiTLSRanges(void** teb) nothrow @nogc
+    {
+    }
+    void scanTLSRanges(void** teb, scope void delegate(void* pbeg, void* pend) nothrow dg) nothrow
```

Comment:
Now added scanning of TLS of multiple DLLs.

The static version of this section could be replaced with the generic one used for `version (Shared)` at the cost of a couple of indirections. That would allow removing bracketing symbols generated here: https://github.com/dlang/dmd/blob/master/compiler/src/dmd/backend/mscoffobj.d#L1201